Add guide for setting up codesigning (#988)
microsoft/AL-Go: Scenarios/Codesigning.md

This guide will take you through how to set up your AL-Go project with an Azure Key Vault and how to use a certificate in the Key Vault to perform code signing. Before you get started, please make sure you've set up your AL-Go project with an Azure Key Vault by following [Scenario 7: Use Azure KeyVault for secrets with AL-Go](./UseAzureKeyVault.md). 

**Note** If your code signing certificate was issued after June 1st 2023 you will most likely need to create a Premium SKU Key Vault. You can [learn more about the differences between Standard and Premium SKU here](https://azure.microsoft.com/en-us/pricing/details/key-vault/)

## Setting up your Azure Key Vault for code signing

> 1. Import your certificate into the Key Vault.

Reviewer: How? do you have any information about how to import a certificate when you don't get a certificate from the provider?

Author: It may depend on which CA the certificate is coming from. Some have Azure Key Vault integrations. Others don't. I can try to compile a list of resources from a couple of CAs on how to do this? Other than that, I'm not sure how much more I can document this process.

2. Configure an Azure Key Vault access policy for the service principal that will be used for signing. At minimum, the account needs the following permissions:

* Cryptographic Operations: Sign
* Certificate Management Operations: Get

> ![Key Vault Access Policies](https://raw.githubusercontent.com/microsoft/AL-Go/main/Scenarios/images/keyvaultaccesspolicies.png)

Reviewer: The way to add an image without taking up space in the repo (slowing down download speed of actions) is to go to https://github.com/microsoft/AL-Go/issues - create a new issue and paste the image there - grab the URL created and cancel the creation of the issue.
Now you have a URL like this https://github.com/microsoft/AL-Go/assets/10775043/f502cec9-1835-4995-8a08-4aa0382fe75f  use that in the .md file instead and don't include the .png file in the PR

# Codesigning in AL-Go

> On June 1, 2023 the industry standards for storing code signing certificates changed. Certificate Authorities now require code signing certificates to be stored on Hardware Security Modules (HSM) or Hardware Tokens that are certified with FIPS 140-2 Level 2 or equivalent. Code signing certificates issued after this date are therefore only issued via physical USB tokens, into on-premises HSM services, or cloud HSM services such as Azure Key Vault. In AL-Go we have decided to opt for using Azure Key Vaults to store code signing certificates and .NET Sign to sign files. 

Reviewer 2: Or change the one below so that the dates are in the same format.

```suggestion
On June 1st 2023 the industry standards for storing code signing certificates changed. Certificate Authorities now require code signing certificates to be stored on Hardware Security Modules (HSM) or Hardware Tokens that are certified with FIPS 140-2 Level 2 or equivalent. Code signing certificates issued after this date are therefore only issued via physical USB tokens, into on-premises HSM services, or cloud HSM services such as Azure Key Vault. In AL-Go we have decided to opt for using Azure Key Vaults to store code signing certificates and .NET Sign to sign files. 
```

How you do this might depend on which Certificate Authority you are getting your certificate from. DigiCert and GlobalSign have integrations with Azure Key Vault. You can follow [this guide](https://learn.microsoft.com/en-us/azure/key-vault/certificates/how-to-integrate-certificate-authority) on how to set up that integration if you are using one of those CAs. Once you have set up the integration, you can request a certificate from within your Azure Key Vault. If you are using another CA you can try following this guide to [Generate a CSR and Install a Certificate in Microsoft Azure Key Vault](https://www.ssl.com/how-to/generate-csr-install-certificate-microsoft-azure-key-vault/). If neither of those options work for you, please engage with your CA to get the certificate into the Key Vault.

2. Configure an Azure Key Vault access policy for the service principal that will be used for signing. At minimum, the account needs the following permissions:

* Cryptographic Operations: Sign
* Certificate Management Operations: Get
![Key Vault Access Policies](https://github.com/microsoft/AL-Go/assets/117829001/c93375e0-ce5b-4aa0-a6b9-a845a87fddef)

## Setting up AL-Go for Code Signing

> Once you have an Azure Key Vault with your certificate in it and a Service Principal with access to the Key Vault you are ready to set up AL-Go for codesigning.

Reviewer 2: ```suggestion
Once you have an Azure Key Vault with your certificate in it and a Service Principal with access to the Key Vault you are ready to set up AL-Go for code signing.
```

1. Update your AL-Go settings with
```json

> "keyVaultCodesignCertificateName": "<Name of your Codesigning certificate>"

Reviewer 2: ```suggestion
"keyVaultCodesignCertificateName": "<Name of your code signing certificate>"
```
